src/tink/core/Promise.hx

```diff
@@ -139,6 +139,18 @@ abstract Promise<T>(Surprise<T, Error>) from Surprise<T, Error> to Surprise<T, E
 
     return loop(0);
   }
+  
+  static public function cache<T>(gen:Void->Promise<T>, expire:Void->Future<Noise>):Void->Promise<T> {
+    var p = null;
+    return function() {
+      if(p == null) {
+        p = gen();
+        expire().handle(function(_) p = null);
+      }
+      #if debug if(p == null) throw 'Having a sync expire function means that you don\'t need to call Promise.cache at all'; #end
+      return p;

```

Comment:
Here I'd suggest: 

```haxe
return p.map(function (o) {
  if (!o.isSuccess()) p = null;
  return o;
});
```

I'd say you don't want to cache failures. In particular since client code may be retrying and then wind up in a synchronous loop.